components/TextFields/src/MDCTextInputController.m

```diff
@@ -0,0 +1,1162 @@
+static NSString *const MDCTextInputControllerInlinePlaceholderColorKey =
+    @"MDCTextInputControllerInlinePlaceholderColorKey";
+static NSString *const MDCTextInputControllerPresentationStyleKey =
+    @"MDCTextInputControllerPresentationStyleKey";
+static NSString *const MDCTextInputControllerTextInputKey = @"MDCTextInputControllerTextInputKey";
+static NSString *const MDCTextInputControllerUnderlineViewModeKey =
+    @"MDCTextInputControllerUnderlineViewModeKey";
+
+static NSString *const MDCTextInputControllerKVOKeyFont = @"font";
+
+static inline CGFloat MDCCeil(CGFloat value) {
+#if CGFLOAT_IS_DOUBLE
+  return ceil(value);
+#else
+  return ceilf(value);
+#endif
+}
+
+static inline CGFloat MDCRound(CGFloat value) {
+#if CGFLOAT_IS_DOUBLE
+  return rint(value);
+#else
+  return rintf(value);
+#endif
+}
+
+static inline UIColor *MDCTextInputInlinePlaceholderTextColor() {
+  return [UIColor colorWithWhite:0 alpha:MDCTextInputHintTextOpacity];
+}
+
+static inline UIColor *MDCTextInputActiveUnderlineColor() {
+  return [MDCPalette indigoPalette].tint500;
+}
+
+static inline UIColor *MDCTextInputNormalUnderlineColor() {
+  return [UIColor lightGrayColor];
+}
+
+static inline UIColor *MDCTextInputTextErrorColor() {
+  return [MDCPalette redPalette].tint500;
+}
+
+@interface MDCTextInputController () {
+  BOOL _mdc_adjustsFontForContentSizeCategory;
+}
+
+@property(nonatomic, strong) NSLayoutConstraint *characterCountY;
+@property(nonatomic, strong) NSLayoutConstraint *characterCountTrailing;
+@property(nonatomic, strong) NSLayoutConstraint *clearButtonY;
+@property(nonatomic, strong) NSLayoutConstraint *clearButtonTrailingCharacterCountLeading;
+@property(nonatomic, strong) UIFont *customLeadingFont;
+@property(nonatomic, strong) UIFont *customPlaceholderFont;
+@property(nonatomic, strong) UIFont *customTrailingFont;
+@property(nonatomic, copy) NSString *errorText;
+@property(nonatomic, copy) NSString *errorAccessibilityValue;
+@property(nonatomic, strong) NSLayoutConstraint *heightConstraint;
+@property(nonatomic, strong) MDCTextInputAllCharactersCounter *internalCharacterCounter;
+@property(nonatomic, readonly) BOOL isDisplayingCharacterCountError;
+@property(nonatomic, readonly) BOOL isDisplayingErrorText;
+@property(nonatomic, readonly) BOOL isPlaceholderUp;
+@property(nonatomic, assign) BOOL isRegisteredForKVO;
+@property(nonatomic, strong) NSArray<NSLayoutConstraint *> *placeholderAnimationConstraints;
+@property(nonatomic, assign) CGRect placeholderDefaultPositionFrame;
+@property(nonatomic, strong) NSLayoutConstraint *placeholderLeading;
+@property(nonatomic, strong) NSLayoutConstraint *placeholderTop;
+@property(nonatomic, strong) NSLayoutConstraint *placeholderTrailingCharacterCountLeading;
+@property(nonatomic, strong) NSLayoutConstraint *placeholderTrailingSuperviewTrailing;
+@property(nonatomic, copy) NSString *previousLeadingText;
+@property(nonatomic, copy) UIColor *previousPlaceholderColor;
```

Comment:
No. The controller needs to know the value of the actual placeholder color and hold it in while the color itself is changed on the placeholder. If it were the same pointer, those mutations would sync back here where we need the state to be temporarily saved.